plots: document html template (#2472)
iterative/dvc.org: content/docs/command-reference/plots/index.md

- `<DVC_METRIC_Y_TITLE>` (optional) - field name to display as the X axis label

## HTML templates

It's possible to supply an HTML file to `dvc plot show` and `dvc plot diff` by

> using the the `--html-template` option. This allows you to customize the
> container where DVC will inject plots it generates.

Reviewer: ⌛  Extracted from the example:

```suggestion
using the the `--html-template` option. This allows you to customize the
container where DVC will inject the plots it generates.
For example, you can add extra content to the web page, like banners or links.
```

> ⚠️ This is a separate feature from
> [custom Vega templates](/doc/command-reference/plots#custom-templates).

The only requirement for this HTML file is to specify the place to inject plots
with a `{plot_divs}` marker. See an

> [example](/doc/command-reference/plots#example-for-the-html-template) that uses

Reviewer: ```suggestion
[example](/doc/command-reference/plots#example-offline-html-template) that uses
```

- `<DVC_METRIC_Y_TITLE>` (optional) - field name to display as the X axis label

## HTML templates

> It's possible to supply a custom HTML file to the `--html-template` option of
> `dvc plot show` and `dvc plot diff` when you would like to customize the _output
> template._

Reviewer: ```suggestion
It's possible to supply an HTML file to
`dvc plot show` and `dvc plot diff` by using the the `--html-template` option.
This allows you to customize the container where DVC will inject plots it generates.
```

> ⚠️ This is a separate feature from
> [custom Vega templates](/doc/command-reference/plots#custom-templates).

> > `--html-template` provides a custom HTML file to inject plots generated by
> > DVC.

Reviewer: ```suggestion
```

The only requirement for this HTML file is to specify the place to inject plots

> with a `{plot_divs}` marker. This way, you can customize the plots output.

Reviewer: ```suggestion
with a `{plot_divs}` marker.
```

> [An example](/doc/command-reference/plots#example-for-the-html-template) that
> uses this feature to provide plots that don't require Internet access is below.

Reviewer: ```suggestion
See an [example](/doc/command-reference/plots#example-for-the-html-template) that
uses this feature to render DVC plots without an Internet connection, below.
```

- `<DVC_METRIC_Y_TITLE>` (optional) - field name to display as the X axis label

## HTML templates

By default, the plots generated by `dvc plots` require Internet access to

> provide Vega JavaScript libraries. It's possible to supply a custom HTML file to
> `dvc plot show` with `--html-template` option when the Internet access is not
> available.

Reviewer: ```suggestion
load Vega JavaScript libraries. It's possible to supply a custom HTML file to
the `--html-template` option of `dvc plot show` and `dvc plot diff` when the
Internet access is not available.
```

> Internet access is not available.

Reviewer: We shouldn't present this as a solution for offline envs initially. You can also use this feature even when you're online. Let's move most of that motivation to the example instead please.

Reviewer: So what's a more general motivation for this section that explains the feature? Perhaps something like "The default plots use a plain HTML file which uses cloud Vega libraries." Plus maybe link to the official "Vega libs" location. Is it https://github.com/vega/vega/releases?

> Internet access is not available.

Reviewer: We shouldn't present this as a solution for offline envs initially. You can also use this feature even when you're online. Let's move most of that motivation to the example instead please.

Reviewer: So what's a more general motivation for this section that explains the feature? Perhaps something like "The default plots use a plain HTML file which uses cloud Vega libraries." Plus maybe link to the official "Vega libs" location. Is it https://github.com/vega/vega/releases?

Reviewer 2: Another motivation may be to put some extra stuff to the plots, like a banner, title etc. But in general, I think people will understand if something works _offline_, it can also work _online._

Reviewer: Yes offline vs online is not the most general use, right? Templates are about customization in general. Offline js imports just happens to be one thing you can customize.

> ⚠️ This is a separate feature from

> > [custom plot templates](/doc/command-reference/plots#custom-templates).
> > `--html-template` allows to provide a custom HTML file to put _generated
> > plots._

Reviewer: ```suggestion
> [custom Vega templates](/doc/command-reference/plots#custom-templates).
> `--html-template` provides a custom HTML file to inject plots generated
> by DVC.
```

> You can create an HTML file with `<script>` references to local Vega libraries.

Reviewer: Same. That's not the sole use of this feature.

Reviewer 2: Main motivation is generating plots without Internet access and (maybe) customizing them. References to _local_ Vega libraries captures both. Without Vega, the user cannot produce plots.

Reviewer: I realize that was the original purpose of the PR but it's just one application of customization, I think.

> Use `{plot_divs}` as a placeholder for DVC to inject the plot images. See

Reviewer: > inject the plot images

Actually is it images exactly what gets injected? Maybe I got this wrong/inexact.

Reviewer 2: Technically they are not the images, plots are produced by Vega library and probably they are drawn on HTML canvas, take user input for modification etc.

> You can create an HTML file with `<script>` references to Vega libraries and put
> `{plot_divs}` as a placeholder for the generated plots. See
> [the example](/doc/command-reference/plots#example-for-the-html-template) for

Reviewer: ```suggestion
You can create an HTML file with `<script>` references to local Vega libraries. Use
`{plot_divs}` as a placeholder for DVC to inject the plot images. See
[this example](/doc/command-reference/plots#example-for-the-html-template) for
```

- `<DVC_METRIC_Y_TITLE>` (optional) - field name to display as the X axis label

## HTML templates

By default, the plots generated by `dvc plots` require Internet access to
provide Vega JavaScript libraries. It's possible to supply a custom HTML file to
`dvc plot show` with `--html-template` option when Internet access is not
available

> ⚠️ This is a separate feature from
> [custom plot templates](/doc/command-reference/plots#custom-templates) to
> provide a custom HTML file for the _generated plots._

You can create an HTML file with `<script>` references to Vega libraries and put
`{plot_divs}` as a placeholder for the generated plots. You can supply this HTML
file with `--html-template` option.

### Example for the custom HTML template

> First, download the Vega libraries and put them to a directory you produce the

Reviewer: This part is too long IMO. And basically the same as the new guide being introduced. Let's not have all this in the cmd desc? Esp. since it's repeated. Can we summarize more? The desc. for now should just explain I think. Eventually (#2572) we'll extract to the UG properly.

Reviewer 2: The guide won't be there. I'll delete the last sentence as it repeats the option.

Reviewer: > I'll delete the UG file if everything looks all right.

I see the intention is to only have the example here. OK how about an actual **Example** instead of making the ref description so long?

Reviewer: p.s. the example text may need a bit of copy editing similar to what I mentioned in https://github.com/iterative/dvc.org/pull/2472#discussion_r656806369. BTW I can help with that if needed.

Reviewer 2: moving example to a section below?

I didn't do to keep it separate from other examples but I can move to the bottom.

Reviewer: Yep, moving it after https://dvc.org/doc/command-reference/plots#example-confusion-matrix basically. To the very bottom of the doc. basically (but easy to find with the right ToV nav) + link from the desc.

Reviewer 2: I did this in b86707b  and 7e966f9.

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC (found in `.dvc/plots/confusion.json`).

### Example: Offline HTML Template

The plots generated by `dvc plots` uses Vega JavaScript libraries, and by
default these load [online resources](https://vega.github.io/vega/usage/#embed).

> There may be times when you need to produce plots without Internet access, or
> want to customize the plots output to put some extra content, like banners or
> extra text. DVC allows to replace the HTML file that contains the final plots.

Reviewer: ```suggestion
There may be times when you need to produce plots without Internet access.
DVC allows to replace the HTML file that contains the final plots.
```

> or want to customize the plots output to put some extra content, like banners or extra text

The example isn't about that. Maybe mention these other examples in the Description subsection? Will leave suggestion ⏳

Reviewer 2: I don't think we can enumerate all use cases of templates in the description? We _implement_ one of the example ideas, the rest can be done in a similar way.

Reviewer: I wasn't suggesting to enumerate every single possible case. My point was that this "Offline HTML Template" example is not about extra HTML content so that sentence just seems unrelated. Could just be removed IMO

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC (found in `.dvc/plots/confusion.json`).

> ### Example for the HTML template

Reviewer: ```suggestion
### Example: Offline HTML Template
```

The plots generated by `dvc plots` uses Vega JavaScript libraries, and by

> default these employ
> [online resources](https://vega.github.io/vega/usage/#embed) to present the
> plots. There may be times when you need to produce plots without Internet

Reviewer: ```suggestion
default these load
[online resources](https://vega.github.io/vega/usage/#embed).
There may be times when you need to produce plots without Internet
```

access, or want to customize the plots output to put some extra content, like
banners or extra text. DVC allows to replace the HTML file that contains the
final plots.

Download the Vega libraries into the directory where you'll produce the

> `dvc plots`.

Reviewer: ```suggestion
`dvc plots`:
```

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC (found in `.dvc/plots/confusion.json`).

### Example for the HTML template

> Download the Vega libraries and put them to a directory you produce the plots.

Reviewer: ```suggestion
Download the Vega libraries into the directory where you'll produce the
`dvc plots`.
```

```dvc
$ wget https://cdn.jsdelivr.net/npm/vega@5.10.0 -O my_vega.js
$ wget https://cdn.jsdelivr.net/npm/vega-lite@4.8.1 -O my_vega_lite.js
$ wget https://cdn.jsdelivr.net/npm/vega-embed@6.5.1 -O my_vega_embed.js
```

Create the following HTML file with your text editor:

```html
<html>
  <head>
    <script src="my_vega.js" type="text/javascript"></script>
    <script src="my_vega_lite.js" type="text/javascript"></script>
    <script src="my_vega_embed.js" type="text/javascript"></script>
  </head>
  <body>
    {plot_divs}
  </body>
</html>
```

Note that this is a standard HTML file with only `{plot_divs}` as a placeholder

> for the generated plots. `<script>` tags in this file point to local JavaScript
> libraries we downloaded above. Supposing this template named as
> `.dvc/page_template.html`, we can then use it like:

Reviewer: ```suggestion
for the generated plots. `<script>` tags in this file point to the local
JavaScript libraries we downloaded above. Supposing this template was
named `.dvc/page_template.html`, we can use it like this:
```

```dvc
$ dvc plots show --html-template .dvc/page_template.html
```

> You can also make it the default html template by setting the configuration
> variable `plots.html_template`.

Reviewer: ```suggestion
You can also make it the default HTML template by setting it as
`dvc config` parameter `plots.html_template`.
```

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC (found in `.dvc/plots/confusion.json`).

### Example for the HTML template

> Download the Vega libraries into the directory where you'll produce the

Reviewer: Should the title be more specific? Something about **offline environments** or **without Internet access**.

Reviewer: (And let's start with a quick motivation intro as mentioned earlier 🙂)

`dvc plots`.

```dvc
$ wget https://cdn.jsdelivr.net/npm/vega@5.10.0 -O my_vega.js
$ wget https://cdn.jsdelivr.net/npm/vega-lite@4.8.1 -O my_vega_lite.js
$ wget https://cdn.jsdelivr.net/npm/vega-embed@6.5.1 -O my_vega_embed.js
```

> Create the following HTML file with your text editor:

Reviewer: Maybe specify the file name for the example here. It's a specific example anyway so it's fine.

Reviewer 2: I mention it later, in sentence starting with "Supposing..."

Reviewer: I'll make a specific suggestion about this part. Resolving

```html
<html>
  <head>
    <script src="my_vega.js" type="text/javascript"></script>
    <script src="my_vega_lite.js" type="text/javascript"></script>
    <script src="my_vega_embed.js" type="text/javascript"></script>
  </head>
  <body>
    {plot_divs}
  </body>
</html>
```

Note that this is a standard HTML file with only `{plot_divs}` as a placeholder

> for the generated plots. `<script>` tags in this file point to the local

Reviewer: How about "for DVC to inject the plots."

Reviewer: Also, the `<scripts>` part is pretty obvious, I think? Maybe mention it in the paragraph before the HTML code block.

JavaScript libraries we downloaded above. Supposing this template was named

> `.dvc/page_template.html`, we can use it like this:

Reviewer: > .dvc/page_template.html

Is that a recommended place to put it? What about just the repo root, or .dvc/tmp/plot/templates/ cc @pared

Reviewer 2: The earlier example by @pared used this path but it's not mandatory.

Reviewer 2: I mention the `<script>` tags a few times because they are the really important part here. They should point to the files we downloaded and that may be a bit tricky.

Reviewer: I would use path `.dvc/plots/` e.g.:

```suggestion
`.dvc/plots/myreport.html`, we can use it like this:
```

Since that's what's used for https://dvc.org/doc/command-reference/plots#custom-templates too.

Reviewer: I'll move this to a specific suggestion too. Resolving

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC (found in `.dvc/plots/confusion.json`).

### Example for the HTML template

There may be times when you need to produce plots without Internet access, or
want to customize the plots output to put some extra content, like banners or
extra text. DVC allows to replace the HTML file that contains the final plots.

Download the Vega libraries into the directory where you'll produce the
`dvc plots`.

```dvc
$ wget https://cdn.jsdelivr.net/npm/vega@5.10.0 -O my_vega.js
$ wget https://cdn.jsdelivr.net/npm/vega-lite@4.8.1 -O my_vega_lite.js
$ wget https://cdn.jsdelivr.net/npm/vega-embed@6.5.1 -O my_vega_embed.js
```

> Create the following HTML file with your text editor:

Reviewer: ```suggestion
Create the following HTML file and save it in `.dvc/plots/mypage.html`:
```

Reviewer: I'm not 100% sure about the location. As long as it doesn't cause confusion with regular plots templates... But if not I'd put it somewhere in the workspace and not in `.dvc/` I think.

Reviewer: > if not I'd put it somewhere in the workspace and not in .dvc/ I think

Except that "the path supplied to `dvc config plots.html_template` is relative to `.dvc/` directory" 🤔 so yeah Idk. Up to you

Reviewer 2: The references in `<script>` tags should also be in the _output_ directory. So if the user wants their plots output in `~/my-project`, the JS files should be there, and keeping this template in `~/my-project/` is a minor nuisance/clutter then. Preferably, these all should be in `~/my-project/plots/` or something like that but this becomes a bit irrelevant to us, so I decided to put everything under project root.

Reviewer: > The references in <script> tags should also be in the output directory

OK so should we use just `./mypage.html` in this example? (updated suggestions)

> I decided to put everything under project root.

I'm confused. Project root is `.`. The example uses `.dvc/plots/`.

> See also https://github.com/iterative/dvc.org/pull/2472#discussion_r664776851

Reviewer 2: Changed as your original suggestion. Thank you.

```html
<html>
  <head>
    <script src="my_vega.js" type="text/javascript"></script>
    <script src="my_vega_lite.js" type="text/javascript"></script>
    <script src="my_vega_embed.js" type="text/javascript"></script>
  </head>
  <body>
    {plot_divs}
  </body>
</html>
```

Note that this is a standard HTML file with only `{plot_divs}` as a placeholder
for DVC to inject plots. `<script>` tags in this file point to the local

> JavaScript libraries we downloaded above. Supposing this template was named
> `.dvc/page_template.html`, we can use it like this:
> 
> ```dvc
> $ dvc plots show --html-template .dvc/page_template.html

Reviewer: ````suggestion
Javascript libraries we downloaded above. We can use it like this:

```dvc
$ dvc plots show --html-template .dvc/plots/mypage.html
````

Reviewer 2: For JavaScript spelling: https://dev.to/gypsydave5/how-to-spell-javascript-107d :))

Reviewer: I didn't mean to change that spelling. Updated suggestion. I assume you're OK with the changes otherwise 🙂

```

You can also make it the default HTML template by setting it as `dvc config`
parameter `plots.html_template`.

```dvc

> $ dvc config plots.html_template page_template.html

Reviewer: ```suggestion
$ dvc config plots.html_template plots/mypage.html
```

Reviewer 2: done as your original suggestion.
